### PR TITLE
Support zIndex Config for Transition

### PR DESCRIPTION
**Pull request by qdsfdhvh**
To implement an animation such as the following, we need to use the zindex value.

<image src="https://github.com/adrielcafe/voyager/assets/17807925/589e79a2-1d96-44bd-bfc8-67c5036d2a78" width="200"/>

### CODE REVIEW

File: voyager-transitions/src/commonMain/kotlin/cafe/adriel/voyager/transitions/ScreenTransition.kt [hunk 1]
```diff
@@ -67,57 +80,33 @@ public fun ScreenTransition(
...
-    content: ScreenTransitionContent = { it.Content() }
-) {
-    ScreenTransition(
-        navigator = navigator,
-        enterTransition = enterTransition,
-        exitTransition = exitTransition,
-        modifier = modifier,
-        content = content
-    )
-}
```
Comment (DevSrSouza) [inline voyager-transitions/src/commonMain/kotlin/cafe/adriel/voyager/transitions/ScreenTransition.kt:85]: Why this was removed? we are keeping it for ABI compatibility

File: voyager-transitions/api/desktop/voyager-transitions.api [hunk 2]
```diff
@@ -62,11 +58,10 @@ public final class cafe/adriel/voyager/transitions/ScaleTransitionKt {
 public final class cafe/adriel/voyager/transitions/ScreenTransition$DefaultImpls {
 	public static fun enter (Lcafe/adriel/voyager/transitions/ScreenTransition;Lcafe/adriel/voyager/core/stack/StackEvent;)Landroidx/compose/animation/EnterTransition;
 	public static fun exit (Lcafe/adriel/voyager/transitions/ScreenTransition;Lcafe/adriel/voyager/core/stack/StackEvent;)Landroidx/compose/animation/ExitTransition;
+	public static fun zIndex (Lcafe/adriel/voyager/transitions/ScreenTransition;Lcafe/adriel/voyager/core/stack/StackEvent;)F
 }
 
 public final class cafe/adriel/voyager/transitions/ScreenTransitionKt {
-	public static final fun ScreenTransition (Lcafe/adriel/voyager/navigator/Navigator;Lkotlin/jvm/functions/Function1;Landroidx/compose/ui/Modifier;Lkotlin/jvm/functions/Function4;Landroidx/compose/runtime/Composer;II)V
-	public static final fun ScreenTransition (Lcafe/adriel/voyager/navigator/Navigator;Lkotlin/jvm/functions/Function1;Lkotlin/jvm/functions/Function1;Landroidx/compose/ui/Modifier;Lkotlin/jvm/functions/Function4;Landroidx/compose/runtime/Composer;II)V
```
Comment (DevSrSouza) [inline voyager-transitions/api/desktop/voyager-transitions.api:69]: Revert this changes, we don't want to break binary compatibility.

File: voyager-transitions/api/android/voyager-transitions.api [hunk 3]
```diff
@@ -62,11 +58,10 @@ public final class cafe/adriel/voyager/transitions/ScaleTransitionKt {
 public final class cafe/adriel/voyager/transitions/ScreenTransition$DefaultImpls {
 	public static fun enter (Lcafe/adriel/voyager/transitions/ScreenTransition;Lcafe/adriel/voyager/core/stack/StackEvent;)Landroidx/compose/animation/EnterTransition;
 	public static fun exit (Lcafe/adriel/voyager/transitions/ScreenTransition;Lcafe/adriel/voyager/core/stack/StackEvent;)Landroidx/compose/animation/ExitTransition;
+	public static fun zIndex (Lcafe/adriel/voyager/transitions/ScreenTransition;Lcafe/adriel/voyager/core/stack/StackEvent;)F
 }
 
 public final class cafe/adriel/voyager/transitions/ScreenTransitionKt {
-	public static final fun ScreenTransition (Lcafe/adriel/voyager/navigator/Navigator;Lkotlin/jvm/functions/Function1;Landroidx/compose/ui/Modifier;Lkotlin/jvm/functions/Function4;Landroidx/compose/runtime/Composer;II)V
-	public static final fun ScreenTransition (Lcafe/adriel/voyager/navigator/Navigator;Lkotlin/jvm/functions/Function1;Lkotlin/jvm/functions/Function1;Landroidx/compose/ui/Modifier;Lkotlin/jvm/functions/Function4;Landroidx/compose/runtime/Composer;II)V
```
Comment (DevSrSouza) [inline voyager-transitions/api/android/voyager-transitions.api:69]: Revert this changes, we don't want to break binary compatibility.

File: voyager-transitions/src/commonMain/kotlin/cafe/adriel/voyager/transitions/ScreenTransition.kt [hunk 4]
```diff
@@ -129,7 +118,9 @@ public fun ScreenTransition(
     navigator: Navigator,
     transition: AnimatedContentTransitionScope<Screen>.() -> ContentTransform,
     modifier: Modifier = Modifier,
+    contentAlignment: Alignment = Alignment.TopStart,
     disposeScreenAfterTransitionEnd: Boolean = false,
+    contentKey: (Screen) -> Any = { it.key },
```
Comment (DevSrSouza) [inline voyager-transitions/src/commonMain/kotlin/cafe/adriel/voyager/transitions/ScreenTransition.kt:123]: What is the reason for `contentKey: (Screen) -> Any = { it.key },` ? Should we expect behavior changes?

File: voyager-transitions/src/commonMain/kotlin/cafe/adriel/voyager/transitions/ScreenTransition.kt [hunk 5]
```diff
@@ -147,8 +138,9 @@ public fun ScreenTransition(
         }
     }
 
-    AnimatedContent(
-        targetState = navigator.lastItem,
+    val contentTransition = updateTransition(targetState = navigator.lastItem, label = "screen_transition")
+    contentTransition.AnimatedContent(
```
Comment (DevSrSouza) [inline voyager-transitions/src/commonMain/kotlin/cafe/adriel/voyager/transitions/ScreenTransition.kt:142]: That is the reason behind this change?

**Review by DevSrSouza — CHANGES_REQUESTED** [review]
It seems that this PR is doing more then just adding zIndex support, can you explaim in depth the changes and why they are needed and the use cases for it?

**Comment by qdsfdhvh** [comment]
I want to support animate like this, with zIndex:

https://github.com/adrielcafe/voyager/assets/17807925/901b2f97-7da2-4fd5-88d5-2de3c8e7208d

without zIndex:

https://github.com/adrielcafe/voyager/assets/17807925/e941fd6c-614c-4fd5-9a1f-b9372cb51532

AnimatedContent By default the targetState content will be at the top, so we need to configure the ZIndex to adjust the hierarchy;

`contentAlignment: Alignment` and `contentKey: (Screen) -> Any` are just the parameters of the AnmiationContent passed out, maybe it will be needed later;

For `contentKey: (Screen) -> Any = { it.key }` is just me that thinks `{ it.key }` might be a bit more appropriate than the default `{it}`，also wo can back to the default value.

File: voyager-transitions/src/commonMain/kotlin/cafe/adriel/voyager/transitions/ScreenTransition.kt [hunk 6]
```diff
@@ -39,6 +39,14 @@ public interface ScreenTransition {
...
      */
     public fun exit(lastEvent: StackEvent): ExitTransition? = null
+
+    /**
+     * Defines the z-index for the Screen.
+     *
+     * @param lastEvent - lastEvent in the navigation stack.
+     * @return Float value for the z-index.
+     */
+    public fun zIndex(lastEvent: StackEvent): Float = 0f
```
Comment (DevSrSouza) [inline voyager-transitions/src/commonMain/kotlin/cafe/adriel/voyager/transitions/ScreenTransition.kt:49]: Can we define zIndex as null? Mostly because the other functions (exit and enter) when null it use the default behavior defined in the `ScreenTransition`, but, this is not the case for `zIndex` here.

**Review by DevSrSouza — APPROVED** [review]

**Comment by DevSrSouza** [comment]
Can you explain what is this change are different from https://github.com/adrielcafe/voyager/pull/445?

**Comment by qdsfdhvh** [comment]
The difference is still the same as enter/exit animation, #455 works for the default animation which needs to configure zIndex, but it's not convenient if it's a screen specific animation which needs to configure zIndex.

**Review by DevNatan — APPROVED** [review]
